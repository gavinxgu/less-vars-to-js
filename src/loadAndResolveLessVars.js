import less from 'less'
import { resolve, dirname } from 'path'
import { readFileSync } from 'fs'
import { getRegexpMatches } from '@hon2a/get-regexp-matches'
import camelCase from 'lodash.camelcase'

const varNameRegExp = /@([\w-]+):/g
const importRegExp = /@import ['"]([^'"]+)['"];/g
const cssVarRegExp = /--([^:]+): ([^;]*);/g

const root = resolve('./')

function loadLessWithImports(entry) {
  const entryPath = resolve('./', entry)
  const input = readFileSync(entryPath, 'utf8')
  const imports = getRegexpMatches(importRegExp, input).map(match => {
    const importPath = match[1]
    const fullImportPath = /\.less$/.test(importPath) ? importPath : `${importPath}.less`
    const resolvedImportPath = /^~/.test(importPath)
      ? resolve(root, 'node_modules', fullImportPath.slice(1))
      : resolve(dirname(entryPath), fullImportPath)
    return {
      match,
      resolved: loadLessWithImports(resolvedImportPath)
    }
  })
  imports.reverse()
  return imports.reduce(
    (acc, { match, resolved }) => acc.substring(0, match.index) + resolved + acc.substring(match[0].length),
    input
  )
}

async function resolveLessVariables(input) {
  const varNames = getRegexpMatches(varNameRegExp, input).map(([, varName]) => varName)
  const { css } = await less.render(
    `${input} #resolved {\n${varNames.map(varName => `--${varName}: @${varName};`).join('\n')}\n}`
  )
  return getRegexpMatches(cssVarRegExp, css.replace(/#resolved {(.*)}/, '$1')).reduce(
    (acc, [, varName, value]) => ({ ...acc, [camelCase(varName)]: value }),
    {}
  )
}

export async function loadAndResolveLessVars(entry) {
  const input = loadLessWithImports(entry)
  return await resolveLessVariables(input)
}
