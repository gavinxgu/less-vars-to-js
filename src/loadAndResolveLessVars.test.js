import { resolve } from 'path'
import { ensureFileSync, removeSync } from 'fs-extra'
import { writeFileSync, existsSync } from 'fs'

import { loadLessWithImports, loadAndResolveLessVars } from './loadAndResolveLessVars'

const dummyExternalDependencyFolder = resolve('./node_modules/dummy')
const dummyExternalDependencyPath = resolve(dummyExternalDependencyFolder, 'variables.less')

beforeAll(() => {
  if (!existsSync(dummyExternalDependencyPath)) {
    ensureFileSync(dummyExternalDependencyPath)
    writeFileSync(dummyExternalDependencyPath, '@base-color: #00ff00;\n@light-color: lighten(@base-color, 10);', {
      encoding: 'utf8'
    })
  }
})

afterAll(() => {
  if (existsSync(dummyExternalDependencyFolder)) {
    removeSync(dummyExternalDependencyFolder)
  }
})

describe('loadLessWithImports', () => {
  it('loads Less files, resolving all imports (transitively)', () => {
    expect(loadLessWithImports('test/variables.less')).toMatchSnapshot()
  })
})

describe('loadAndResolveLessVars', () => {
  it('loads Less files, resolves imports, and resolves and collects variables', async () => {
    const output = await loadAndResolveLessVars('test/variables.less')
    expect(output).toEqual({
      baseColor: '#00ff00', // from dummy lib (transitive external import)
      lightColor: '#4dff4d', // original from from dummy lib overwritten locally, computed from imported variable
      primaryColor: 'indigo', // from local import
      darkColor: '#00cc00', // from local import, computed from transitively imported variable
      secondaryColor: 'indigo', // from entry point, simple assignment from imported variable
      errorColor: 'darkred' // from entry point
    })
  })
})
