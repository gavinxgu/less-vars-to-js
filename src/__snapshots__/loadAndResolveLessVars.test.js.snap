// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadLessWithImports loads Less files, resolving all imports (transitively) 1`] = `
"// random comment to have import not on at the very start of the file
@base-color: #00ff00;
@light-color  : lighten(@base-color, 10);

@primary-color: indigo;
@dark-color: darken(@base-color, 10);


@secondary-color: @primary-color;
@error-color: darkred;
@light-color: lighten(@base-color, 15);
"
`;
